rust/sql/src/postgres/reader.rs

```diff
@@ -0,0 +1,894 @@
+fn row_to_schema(row: &postgres::Row) -> Result<Schema> {
+    let mut metadata = std::collections::HashMap::new();
+    let fields = row
+        .columns()
+        .iter()
+        .map(|col: &postgres::Column| {
+            metadata.insert(col.name().to_string(), col.type_().to_string());
+            Field::new(col.name(), pg_to_arrow_type(col.type_()).unwrap(), true)
+        })
+        .collect();
+    Ok(Schema::new_with_metadata(fields, metadata))
+}
+
+fn read_from_binary<R>(mut reader: R, schema: &Schema) -> Result<RecordBatch>
+where
+    R: Read,
+{
+    // read signature
+    let mut bytes = [0u8; 11];
+    reader.read_exact(&mut bytes).unwrap();
+    if bytes != MAGIC {
+        eprintln!("Unexpected binary format type");
+        return Err(ArrowError::IoError(
+            "Unexpected Postgres binary type".to_string(),
+        ));
+    }
+    // read flags
+    let mut bytes: [u8; 4] = [0; 4];
+    reader.read_exact(&mut bytes).unwrap();
+    let _size = u32::from_be_bytes(bytes);
+
+    // header extension area length
+    let mut bytes: [u8; 4] = [0; 4];
+    reader.read_exact(&mut bytes).unwrap();
+    let _size = u32::from_be_bytes(bytes);
+
+    read_rows(&mut reader, schema)
+}
+
+/// Read row tuples
+fn read_rows<R>(mut reader: R, schema: &Schema) -> Result<RecordBatch>
+where
+    R: Read,
+{
+    let mut is_done = false;
+    let field_len = schema.fields().len();
+
+    let mut buffers: Vec<Vec<u8>> = vec![vec![]; field_len];
+    let mut null_buffers: Vec<Vec<bool>> = vec![vec![]; field_len];
+    let mut offset_buffers: Vec<Vec<i32>> = vec![vec![]; field_len];
+
+    let default_values: Vec<Vec<u8>> = schema
+        .fields()
+        .iter()
+        .map(|f| match f.data_type() {
+            DataType::Null => vec![],
+            DataType::Boolean => vec![0],
+            DataType::Int8 => vec![0],
+            DataType::Int16 => vec![0; 2],
+            DataType::Int32 => vec![0; 4],
+            DataType::Int64 => vec![0; 8],
+            DataType::UInt8 => vec![0],
+            DataType::UInt16 => vec![0; 2],
+            DataType::UInt32 => vec![0; 4],
+            DataType::UInt64 => vec![0; 8],
+            DataType::Float16 => vec![0; 2],
+            DataType::Float32 => vec![0; 4],
+            DataType::Float64 => vec![0; 8],
+            DataType::Timestamp(_, _) => vec![0; 8],
+            DataType::Date32(_) => vec![0; 4],
+            DataType::Date64(_) => vec![0; 8],
+            DataType::Time32(_) => vec![0; 4],
+            DataType::Time64(_) => vec![0; 8],
+            DataType::Duration(_) => vec![0; 8],
+            DataType::Interval(_) => vec![0; 8],
+            DataType::Binary => vec![],
+            DataType::FixedSizeBinary(len) => vec![0; *len as usize],
+            DataType::Utf8 => vec![],
+            DataType::List(_) => vec![],
+            DataType::FixedSizeList(_, len) => vec![0; *len as usize],
+            DataType::Struct(_) => vec![],
```

Comment:
These empty `vec![]`s are not necessarily correct/complete. I haven't touched this code in months, so I can't recall how far I got here. Some types are also unreachable due to types not being supported.